src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs

```diff
@@ -0,0 +1,171 @@
+        /// </summary>
+        /// <param name="propertyInfo">The <see cref="PropertyInfo"/> to create a
+        /// <see cref="TagHelperUsageDescriptor"/> from.</param>
+        /// <returns>A <see cref="TagHelperUsageDescriptor"/> that describes the summary and remarks XML documentation
+        /// for the given <paramref name="propertyInfo"/>.</returns>
+        public static TagHelperUsageDescriptor CreateDescriptor([NotNull] PropertyInfo propertyInfo)
+        {
+            var id = XmlDocumentationProvider.GetId(propertyInfo);
+            var declaringAssembly = propertyInfo.DeclaringType.Assembly;
+
+            return CreateDescriptorCore(declaringAssembly, id);
+        }
+
+        private static TagHelperUsageDescriptor CreateDescriptorCore(Assembly assembly, string id)
+        {
+            var assemblyLocation = assembly.Location;
+
+            if (string.IsNullOrEmpty(assemblyLocation) && !string.IsNullOrEmpty(assembly.CodeBase))
+            {
+                var uri = new UriBuilder(assembly.CodeBase);
+
+                // Normalize the path to a UNC path. This will remove things like file:// from start of the uri.Path.
+                assemblyLocation = Uri.UnescapeDataString(uri.Path);
+            }
+
+            // Couldn't resolve a valid assemblyLocation.
+            if (string.IsNullOrWhiteSpace(assemblyLocation))
+            {
+                return null;
+            }
+
+            var xmlDocumentationFile = GetXmlDocumentationFile(assembly, assemblyLocation);
+
+            // Only want to process the file if it exists.
+            if (xmlDocumentationFile != null)
+            {
+                var documentationProvider = new XmlDocumentationProvider(xmlDocumentationFile.FullName);
+
+                var summary = documentationProvider.GetSummary(id);
+                var remarks = documentationProvider.GetRemarks(id);
+
+                if (!string.IsNullOrEmpty(summary) || !string.IsNullOrEmpty(remarks))
+                {
+                    return new TagHelperUsageDescriptor(summary, remarks);
+                }
+            }
+
+            return null;
+        }
+
+        private static FileInfo GetXmlDocumentationFile(Assembly assembly, string assemblyLocation)
+        {
+            try
+            {
+                var assemblyDirectory = Path.GetDirectoryName(assemblyLocation);
+                var assemblyName = Path.GetFileName(assemblyLocation);
+                var assemblyXmlDocumentationName = Path.ChangeExtension(assemblyName, ".xml");
+
+                // Check for a localized XML file for the current culture.
+                var localizedXmlFile = GetLocalizedXmlDocumentationFile(
+                    CultureInfo.CurrentCulture,
+                    assemblyDirectory,
+                    assemblyXmlDocumentationName);
+                FileInfo assemblyXmlDocumentationFile;
```

Comment:
Confusing to have two variables of the same type here; just need one `xmlDocumentationFile` variable.